[CI] CancellableFanOutTests testConcurrency failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/4gldhpe6ejvjo/tests/:server:test/org.elasticsearch.action.support.CancellableFanOutTests/testConcurrency

**Reproduction line:**
```
./gradlew ':server:test' --tests "org.elasticsearch.action.support.CancellableFanOutTests.testConcurrency" -Dtests.seed=1C4A118E452A6BB3 -Dtests.locale=es-AR -Dtests.timezone=Asia/Riyadh -Druntime.java=20
```

**Applicable branches:**
main

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.action.support.CancellableFanOutTests&tests.test=testConcurrency

**Failure excerpt:**
```
java.lang.AssertionError: 
Expected: a string starting with "processor-thread-"
     but: was "TEST-CancellableFanOutTests.testConcurrency-seed#[1C4A118E452A6BB3]"

  at __randomizedtesting.SeedInfo.seed([1C4A118E452A6BB3:BABAD31324798B7A]:0)
  at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
  at org.junit.Assert.assertThat(Assert.java:956)
  at org.junit.Assert.assertThat(Assert.java:923)
  at org.elasticsearch.action.support.CancellableFanOutTests$3.onCompletion(CancellableFanOutTests.java:206)
  at org.elasticsearch.action.support.CancellableFanOutTests$3.onCompletion(CancellableFanOutTests.java:180)
  at org.elasticsearch.action.ActionListener.completeWith(ActionListener.java:313)
  at org.elasticsearch.action.support.CancellableFanOut.lambda$run$0(CancellableFanOut.java:61)
  at org.elasticsearch.common.util.concurrent.RunOnce.run(RunOnce.java:28)
  at org.elasticsearch.action.support.CancellableFanOut.lambda$run$2(CancellableFanOut.java:76)
  at org.elasticsearch.core.AbstractRefCounted$1.closeInternal(AbstractRefCounted.java:117)
  at org.elasticsearch.core.AbstractRefCounted.decRef(AbstractRefCounted.java:69)
  at org.elasticsearch.action.support.RefCountingRunnable.close(RefCountingRunnable.java:116)
  at org.elasticsearch.action.support.CancellableFanOut.run(CancellableFanOut.java:119)
  at org.elasticsearch.action.support.CancellableFanOutTests.testConcurrency(CancellableFanOutTests.java:213)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
  at java.lang.reflect.Method.invoke(Method.java:578)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1623)

```

--- a/server/src/main/java/org/elasticsearch/action/support/CancellableFanOut.java
+++ b/server/src/main/java/org/elasticsearch/action/support/CancellableFanOut.java
@@ -151,8 +151,8 @@ public abstract class CancellableFanOut<Item, ItemResponse, FinalResponse> {
     protected abstract void onItemFailure(Item item, Exception e);
 
     /**
-     * Called when responses for all items have been processed, on the thread that processed the last per-item response. Not called if the
-     * task is cancelled.
+     * Called when responses for all items have been processed, on the thread that processed the last per-item response or possibly the
+     * thread which called {@link #run} if all items were processed before {@link #run} returns. Not called if the task is cancelled.
      * <p>
      * Note that it's easy to accidentally capture another reference to this class when implementing this method, and that will prevent the
      * early release of any accumulated results. Beware of lambdas, and test carefully.

--- a/server/src/test/java/org/elasticsearch/action/support/CancellableFanOutTests.java
+++ b/server/src/test/java/org/elasticsearch/action/support/CancellableFanOutTests.java
@@ -17,6 +17,7 @@ import org.elasticsearch.tasks.TaskCancelledException;
 import org.elasticsearch.tasks.TaskId;
 import org.elasticsearch.test.ESTestCase;
 import org.elasticsearch.test.ReachabilityChecker;
+import org.hamcrest.Matcher;
 
 import java.util.HashMap;
 import java.util.List;
@@ -143,10 +144,16 @@ public class CancellableFanOutTests extends ESTestCase {
         expectThrows(TaskCancelledException.class, future::actionGet);
     }
 
+    private static void assertCurrentThread(Matcher<String> matcher) {
+        assertThat(Thread.currentThread().getName(), matcher);
+    }
+
     public void testConcurrency() throws InterruptedException {
 
         final var isProcessorThread = startsWith("processor-thread-");
         final var isCancelThread = equalTo("cancel-thread");
+        final var isTestThread = startsWith("TEST-");
+        assertCurrentThread(isTestThread);
 
         final var items = randomList(1, 3, () -> randomAlphaOfLength(5));
         final var processorThreads = new Thread[items.size()];
@@ -155,7 +162,7 @@ public class CancellableFanOutTests extends ESTestCase {
         for (int i = 0; i < processorThreads.length; i++) {
             processorThreads[i] = new Thread(() -> {
                 try {
-                    assertThat(Thread.currentThread().getName(), isProcessorThread);
+                    assertCurrentThread(isProcessorThread);
                     final var item = Objects.requireNonNull(queue.poll(10, TimeUnit.SECONDS));
                     safeAwait(barrier);
                     item.run();
@@ -169,7 +176,7 @@ public class CancellableFanOutTests extends ESTestCase {
         final var task = new CancellableTask(1, "test", "test", "", TaskId.EMPTY_TASK_ID, Map.of());
 
         final var cancelThread = new Thread(() -> {
-            assertThat(Thread.currentThread().getName(), isCancelThread);
+            assertCurrentThread(isCancelThread);
             safeAwait(barrier);
             TaskCancelHelper.cancel(task, "test");
         }, "cancel-thread");
@@ -187,14 +194,14 @@ public class CancellableFanOutTests extends ESTestCase {
 
             @Override
             protected void onItemResponse(String s, String response) {
-                assertThat(Thread.currentThread().getName(), isProcessorThread);
+                assertCurrentThread(isProcessorThread);
                 assertEquals(s, response);
                 assertThat(itemsProcessed.incrementAndGet(), lessThanOrEqualTo(items.size()));
             }
 
             @Override
             protected void onItemFailure(String s, Exception e) {
-                assertThat(Thread.currentThread().getName(), isProcessorThread);
+                assertCurrentThread(isProcessorThread);
                 assertThat(e, instanceOf(ElasticsearchException.class));
                 assertEquals("sendItemRequest", e.getMessage());
                 assertThat(itemsProcessed.incrementAndGet(), lessThanOrEqualTo(items.size()));
@@ -203,7 +210,7 @@ public class CancellableFanOutTests extends ESTestCase {
             @Override
             protected String onCompletion() {
                 assertEquals(items.size(), itemsProcessed.get());
-                assertThat(Thread.currentThread().getName(), isProcessorThread);
+                assertCurrentThread(anyOf(isTestThread, isProcessorThread));
                 if (randomBoolean()) {
                     return "finished";
                 } else {
@@ -214,7 +221,7 @@ public class CancellableFanOutTests extends ESTestCase {
             @Override
             public void onResponse(String s) {
                 assertEquals(items.size(), itemsProcessed.get());
-                assertThat(Thread.currentThread().getName(), isProcessorThread);
+                assertCurrentThread(anyOf(isTestThread, isProcessorThread));
                 assertEquals("finished", s);
                 completionLatch.countDown();
             }
@@ -222,10 +229,10 @@ public class CancellableFanOutTests extends ESTestCase {
             @Override
             public void onFailure(Exception e) {
                 if (e instanceof TaskCancelledException) {
-                    assertThat(Thread.currentThread().getName(), anyOf(isProcessorThread, isCancelThread));
+                    assertCurrentThread(anyOf(isTestThread, isProcessorThread, isCancelThread));
                 } else {
                     assertEquals(items.size(), itemsProcessed.get());
-                    assertThat(Thread.currentThread().getName(), isProcessorThread);
+                    assertCurrentThread(anyOf(isTestThread, isProcessorThread));
                     assertThat(e, instanceOf(ElasticsearchException.class));
                     assertEquals("onCompletion", e.getMessage());
                 }

